keps/sig-network/2091-admin-network-policy/README.md

```diff
@@ -0,0 +1,1116 @@
+while keeping the mental model easy to understand and use.
+The focus of this KEP is on cluster scoped controls for east-west traffic within
+a cluster. Cluster scoped controls for north-south traffic may be addressed via
+future versions of the api resources introduced in this or other future KEPs.
+For the time being, the AdminNetworkPolicy resource introduced by this KEP will
+never affect north-south traffic, and thus also don't override or bypass NetworkPolicies
+with ipBlock rules that select external traffic.
+
+## Proposal
+
+In order to achieve the three primary broad use cases for a cluster admin to
+secure K8s clusters, we propose to introduce the following resource
+under `policy.networking.k8s.io` API group:
+- AdminNetworkPolicy
+
+### AdminNetworkPolicy resource
+
+An AdminNetworkPolicy (ANP) resource will help the administrators:
+1. Set strict security rules for the cluster, i.e. a developer CANNOT override
+these rules by creating NetworkPolicies that applies to the same workloads as the
+AdminNetworkPolicy does.
+2. Set baseline security rules that describes default connectivity for cluster
+workloads, which CAN be overridden by developer NetworkPolicies if needed.
+
+#### Actions
+
+Unlike the NetworkPolicy resource in which each rule represents an allowed
+traffic, AdminNetworkPolicy will enable administrators to set `Pass`,
+`Deny` or `Allow` as the action of each rule. AdminNetworkPolicy rules should
+be read as-is, i.e. there will not be any implicit isolation effects for the Pods
+selected by the AdminNetworkPolicy, as opposed to what NetworkPolicy rules imply.
+
+- Pass: Traffic that matches a `Pass` rule will skip all further rules from all 
+  ANPs and instead be enforced by the K8s NetworkPolicies. If there is no K8s 
+  NetworkPolicy rule match, traffic will be allowed.
+- Deny: Traffic that matches a `Deny` rule will be dropped. 
+- Allow: Traffic that matches an `Allow` rule will be allowed.
+
+AdminNetworkPolicy `Deny` rules are useful for administrators to explicitly
+block traffic with malicious clients, or workloads that poses security risks.
+Those traffic restrictions can only be lifted once the `Deny` rules are deleted
+or modified by the admin.
+
+On the other hand, the `Allow` rules can be used to call out traffic in the cluster
+that needs to be allowed for certain components to work as expected (egress to
+CoreDNS for example). Those traffic should not be blocked when developers apply
+NetworkPolicy to their Namespaces which isolates the workloads.
+
+AdminNetworkPolicy `Pass` rules allows an admin to delegate security posture for
+certain traffic to the Namespace owners by overriding any lower priority Allow or 
+Deny rules. For example, allowing intra-tenant traffic can be delegated to tenant 
+admins explicitly by the cluster admin with the use of `Pass` rules.
+
+#### Priority
+
+The policy instances will be ordered based on the numeric priority assigned to the
```

Comment:
Yep you're right, however, we spoke internally, and with @thockin , and ended up trying to keep it as straightforward (in the group's opinion) as possible i.e Higher # == Higher Priority

Reply:
Also here's that convo :) https://github.com/kubernetes/enhancements/pull/2522#discussion_r770733497

Reply:
> ended up trying to keep it as straightforward (in the group's opinion) as possible i.e Higher # == Higher Priority

Gotcha. This feels rather backwards to me but it's probably because I'm familiar with Calico's ordering model, which has always used a lower number for this field to mean a higher priority.

I'll need to think about how to reconcile this if we're dead-set on this - it might prove difficult for Calico users who have already built out paradigms for ordering policies one way to then integrate another API that does ordering in the opposite way :sweat_smile:

Reply:
I liked "order" but we flipped it from "0 is highest importance" to "0 is lowest since the last draft, which is what we discussed at sig-net (with notable objections)

Reply:
> which is what we discussed at sig-net (with notable objections)

I must have missed this meeting... or fallen asleep or something because I have no memory of this conversation :smile:

Reply:
Given the annoying magic around go empty types, I think a useful mode would "regular rules are 1-1000, lower values are more important, 0 is last or disallowed"

Reply:
My preference would be to have a model similar to "regular rules be 1-1000, lower values are more important, 0 is disallowed."

The problem that I still see is that overloading this to mean both "the order of this policy compared to other ANPs" as well as "whether or not this is executed before or after namespaced NetworkPolicies" makes this a complicated field to use.

Did we ever consider something like this?

```go
// 1-1000, zero is reserved, lower is more important.
Order *uint32

// Set to true if this policy should be executed after namespaced policies.
AfterNamespacedPolicies bool
```

Reply:
@squeed mentioned it elsewhere, but the other option I see is to have a middle priority for NPs so that at least the ordering makes intuitive sense when you lay out all of the policies together. It's still not great since "500" in this example is a magic number and I suspect some organizations are going to want more than 500 orders to work with. 

```go

// 1-499: Implemented before NP
// 500: Reserved for Namespaced network policies
// 501-1000: Implemented after NP
Order *int32
```

Reply:
Right, I don't like the magic number either.

Goofy idea: What if this is signed? 0 is reserved for NetworkPolicy. I'll let everyone else argue whether -1000 or 1000 should be executed first :-)